src/metric_family.cc

```diff
@@ -0,0 +1,230 @@
+// Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "metric_family.h"
+
+#ifdef TRITON_PB_STUB
+#include "pb_stub.h"
+#endif
+
+namespace triton { namespace backend { namespace python {
+
+MetricFamily::MetricFamily(
+    const std::string& name, const std::string& description,
+    const MetricKind& kind)
+    : name_(name), description_(description), kind_(kind),
+      metric_family_request_kind_(MetricFamilyNew),
+      metric_family_address_(nullptr)
+{
+#ifdef TRITON_PB_STUB
+  // Send the request to create the MetricFamily to the parent process
+  std::unique_ptr<Stub>& stub = Stub::GetOrCreateInstance();
+  SaveToSharedMemory(stub->ShmPool());
+  CustomMetricsMessage* custom_metrics_msg = nullptr;
+  try {
+    stub->SendCustomMetricsMessage(
+        &custom_metrics_msg, PYTHONSTUB_MetricFamilyRequest, shm_handle_);
+  }
+  catch (const PythonBackendException& pb_exception) {
+    throw PythonBackendException(
+        "Error when creating MetricFamily: " +
+        std::string(pb_exception.what()));
+  }
+  metric_family_address_ = custom_metrics_msg->address;
+#endif
+}
+
+MetricFamily::~MetricFamily()
+{
+#ifdef TRITON_PB_STUB
+  // Send the request to delete the MetricFamily to the parent process
+  if (metric_family_request_kind_ != MetricFamilyDelete) {
+    // Clear all the metrics first
+    for (auto& m : metric_map_) {
+      m.second->Clear();
+    }
```

Comment:
does this need to be protected by `metric_map_mu_` ?

Reply:
Good catch! Fixed.